### DISCUSSION TITLE
Reverse image search implementation

### DISCUSSION THREAD

**Discussion by FoxxMD** [rr-/szurubooru#500] [Q&A]
I think this is a question mostly for @sgsunder 

I have a working "find image based on perceptual hash" system in a project I am working on. It also does fuzzy searching for **in-memory** hashes.

Soon, I want to expand this to storing the hashes so I can search them later. I can do this, now, without an issue but only if I'm searching for **exact** hashes using simple `where` clauses in my main database.

I've been a long time szuru user and I remember when the project migrated from using elasticsearch (with i assume [fuzzy query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-fuzzy-query.html)) to no dependencies!

From looking at the code now my rough understanding of how reverse image search is working:

* using [image-match](https://github.com/ProvenanceLabs/image-match)
* storing some kind of [matrix generated from the image-match signature](https://github.com/rr-/szurubooru/blob/6c3b50d287a8b8ba67d2bb4cf3cffe25bcd3b4e6/server/szurubooru/func/image_hash.py#L228) in the database
* using a [special type of search](https://github.com/rr-/szurubooru/blob/master/server/szurubooru/func/posts.py#L930) (postgres only?) to search for these matrixes in the db

___

I'm trying to avoid making my project dependent on elasticsearch, if possible, so what you've done here is super relevant.

My questions:

* Can you correct/clarify what the "word" matrix for the signature is? And any other implementation details you think are important here
* Is the datatype used to store the signature data special to postgres?
* How has performance panned out with this implementation?
* Any other "gotchas" you can think of -- in general for reverse image searching or this particular implementation?
* If you could do the reverse image search implementation again, would you use the same approach?

Thanks!

**Comment by liamw1** [comment]
Throwing this out there in case it's not completely irrelevant to you now: I've re-implemented the reverse-search algorithm on my fork and I think I understand it pretty well. Were you still wanting to know the answers for any of these questions?
